File: conv/calc/fft.py
import numpy as np

# def convfft(x, w):
#     w.reverse()
#     M = len(x)
#     N = len(w)
#     MN = M + N - 1
#     L = 2 ** (int(np.log2(MN)) + 1)
#     xfft = np.fft.fft(x, L)
#     wfft = np.fft.fft(w, L)
#     yfft = xfft * wfft
#     y = np.fft.ifft(yfft).real[:MN]
#     return y

# if __name__ == '__main__':
#     x = [0, 1, 2, 3, 4, 5]
#     w = [0, 1, 2]
#     print(convfft(x,w))

import numpy as np

def convfft(x, w):
    shape = np.array(x.shape) + np.array(w.shape) - 1
    fsize = 2 ** (np.ceil(np.log2(shape)).astype(int))
    # print(np.rot90(w, 2))
    B = np.fft.fft2(np.rot90(w, 2), fsize)
    A = np.fft.fft2(x, fsize)
    AB = np.fft.ifft2(A * B).real[:shape[0], :shape[1]]
    return AB

if __name__ == '__main__':
    x = np.array([[0, 1, 2], [1, 2, 3], [2, 3, 4]])
    w = np.array([[0, 1], [1, 2]])
    print(convfft(x,w))

File: conv/calc/fft.cpp
#include <iostream>
#include <cmath>

using namespace std;

// void convfft(float* x, float* w, )

int main() {
    float x[] = {0,1,2,1,2,3,3,4,5};
    float w[] = {0,1,1,2};
    int M1 = 3, M2 = 3, N1 = 2, N2 = 2;
    int MN1 = M1 + N1 - 1;
    int MN2 = M2 + N2 - 1;
    int L1 = pow(2, ceil(log2(MN1)));
    int L2 = pow(2, ceil(log2(MN2)));
    float w_trans[] = {0,0,0,0};
    for (int i = 0; i < N1; ++i) {
        for (int j = 0; j < N2; ++j) {
            w_trans[i * N2 + j] = w[(N1 - 1 - i) * N2 + (N2 - 1 - j)];
        }
    }
    // for (int i = 0; i < N1 * N2; ++i) {
    //     printf("%f ", w_trans[i]);
    // }
    printf("\n--%d:%d---\n", L1, L2);
    cout << "helloworld" << endl;
    return 0;
}